ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdRecalculationService.java

```diff
@@ -57,21 +60,27 @@ public void recalculatePersistedInstanceId(String collectionName, int numberOfIn
 		if (!indexDropped) {
 			log.info("Step 1/3\tNo index to drop");
 		}
-		AtomicInteger count = new AtomicInteger();
-		collection.findByQuery(query(where(DOCUMENT_ROUTING_KEY).exists(true)))
-				.peek(it -> {
-					int currentCount = count.incrementAndGet();
-					if (currentCount % 10_000 == 0) {
-						log.info("Step 2/3\tUpdated persisted instance id for {} documents", currentCount);
-					}
-				})
-				.forEach(document -> {
-					Object id = document.get("_id");
-					int instanceId = getInstanceId(document.get(DOCUMENT_ROUTING_KEY), numberOfInstances);
-					collection.updateById(id, Map.of(DOCUMENT_INSTANCE_ID, instanceId));
-				});
 
-		log.info("Step 2/3\tUpdated persisted instance id for {} documents", count.get());
+		try (Stream<List<Document>> batches = collection.findByQuery(query(where(DOCUMENT_ROUTING_KEY).exists(true)), BATCH_SIZE)) {
+			AtomicInteger count = new AtomicInteger();
+			batches.forEach(batch -> {
+				List<Document> updates = batch.stream()
+						.peek(it -> {
+							int currentCount = count.incrementAndGet();
+							if (currentCount % 10_000 == 0) {
+								log.info("Step 2/3\tUpdated persisted instance id for {} documents", currentCount);
+							}
+						})
+						.map(it -> {
+							Object id = it.get("_id");
+							int instanceId = getInstanceId(it.get(DOCUMENT_ROUTING_KEY), numberOfInstances);
```

Comment:
Do we need a safeguard here that `numberOfInstances` (given by the caller of this jmx operation) is equal to the number of running partitions? Otherwise, could there be a problem with documents being written to the db by the mirror here while this operation is running?

Reply:
Yeah, there is no safeguard now. Maybe it is best to just determine the current number of partitions and use that, removing the parameter.

Reply:
I have updated this service now so that it recalculates the instance id for all collections belonging to mirrored types with this feature enabled. I have also changed it so that it determines the number of partitions from the space property `mirror-service.cluster.partitions`, falling back to the system property `cluster.partitions`.
These two changes removes both parameters from the MBean operation.

This was done in https://github.com/AvanzaBank/ymer/pull/62/commits/1ec34df05b928fd8e0726c2202995993456989cd